src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilder.java

```diff
@@ -551,67 +529,33 @@ private static final Pair range(Coordinate[] points, int offset, int length) {
          * @return Array of edges
          */
         protected static Edge[] ring(int component, boolean direction, boolean handedness, BaseLineStringBuilder<?> shell,
-                                     Coordinate[] points, int offset, Edge[] edges, int toffset, int length) {
+                                     GeoPoint[] points, int offset, Edge[] edges, int toffset, int length) {
             // calculate the direction of the points:
-            // find the point a the top of the set and check its
-            // neighbors orientation. So direction is equivalent
-            // to clockwise/counterclockwise
-            final int top = top(points, offset, length);
-            final int prev = (offset + ((top + length - 1) % length));
-            final int next = (offset + ((top + 1) % length));
-            boolean orientation = points[offset + prev].x > points[offset + next].x;
-
-            // OGC requires shell as ccw (Right-Handedness) and holes as cw (Left-Handedness) 
-            // since GeoJSON doesn't specify (and doesn't need to) GEO core will assume OGC standards
-            // thus if orientation is computed as cw, the logic will translate points across dateline
-            // and convert to a right handed system
-
-            // compute the bounding box and calculate range
-            Pair<Pair, Pair> range = range(points, offset, length);
-            final double rng = (Double)range.getLeft().getRight() - (Double)range.getLeft().getLeft();
-            // translate the points if the following is true
-            //   1.  shell orientation is cw and range is greater than a hemisphere (180 degrees) but not spanning 2 hemispheres 
-            //       (translation would result in a collapsed poly)
-            //   2.  the shell of the candidate hole has been translated (to preserve the coordinate system)
-            boolean incorrectOrientation = component == 0 && handedness != orientation;
-            if ( (incorrectOrientation && (rng > DATELINE && rng != 2*DATELINE)) || (shell.translated && component != 0)) {
-                translate(points);
-                // flip the translation bit if the shell is being translated
+            boolean orientation = GeoUtils.computePolyOrientation(points, length, offset);
+            boolean corrected = GeoUtils.correctPolyAmbiguity(points, handedness, orientation, component, offset, length, shell.translated);
+
+            // correct the orientation post translation (ccw for shell, cw for holes)
+            if (corrected && (component == 0 || (component != 0 && handedness == orientation))) {
```

Comment:
Is the `component != 0` check needed? Seems like the only way to get past the `||` is for that condition to be true.